### PR TITLE
Implement compile time derivative framework

### LINKED ISSUE #21157: Explore a replacement for `ExpressionBuilder` that is fully compile time only
## Reason
For some applications we need to compute the derivatives of large expressions and evaluate them separate from the underived expression. We currently have `ExpressionBuilder` which builds an AST using operator overloading at _runtime_, and then uses FParser and JIT for symbolic automatic differentiation and the generation of evaluable functions and their derivatives. However this machinery takes a lot of round trips

1. C++ code using overloaded operators
2. AST generation at runtime
3. serialization of the AST into an FParser expression string
4. Parsing of the string into an FParser AST
5. Conversion of the FParser AST into FParser byte code
6. Conversion of the FParser byte code into C++ (at runtime)
7. Spawning of a compiler command to build an object file
8. using dlopen to load said object file for evaluation

## Design
Let's see if we can fix this using template meta-programming. Expression templates would allow us to build an expression tree and it's derivatives at compile time. Modern compilers should optimize this to the point where the resulting machine code of an expression template and a vanilla expression are identical (see example below)

image

## Impact
Robust alternative for compile time construction of expression derivatives.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #21157

### CODE REVIEW

**Comment by dschwen** [comment]
Here's an example how this is used:

https://github.com/idaholab/moose/pull/21158/files?diff=unified&w=1#diff-211289645810e3645c7d32c0714bff5a57d242f8693e056f7f550c081f94e6d1R22-R55

**Comment by GiudGiud** [comment]
Where is this going?
Is this to support the AD system? If there is someone else knowledgeable about what this is for please review request them

**Comment by GiudGiud** [comment]
Where is this going?
Is this to support the AD system? If there is someone else knowledgeable about what this is for please review request them

**Comment by loganharbour** [comment]
I can review. This is fun and exciting

File: framework/include/utils/CompileTimeDerivatives.h [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+class CTArrayRef : public CTBase
+{
+public:
+  CTArrayRef(const T & arr, const I & idx) : _arr(arr), _idx(idx) {}
+  auto operator()() const { return _arr[_idx]; }
+
+  template <CTTag dtag>
+  auto D() const
+  {
+    if constexpr (tag == dtag)
```
Comment (loganharbour) [inline framework/include/utils/CompileTimeDerivatives.h:189]: 😎

**Review by loganharbour — APPROVED** [review]
That coverage though

**Pull request merged**